storage/namespace.go

```diff
@@ -728,6 +732,51 @@ func (n *dbNamespace) Flush(
 	return res
 }
 
+func (n *dbNamespace) Snapshot(blockStart, callStart time.Time, flush persist.Flush) error {
+	n.RLock()
+	if n.bs != bootstrapped {
+		n.RUnlock()
+		n.metrics.snapshot.ReportError(n.nowFn().Sub(callStart))
+		return errNamespaceNotBootstrapped
+	}
+	n.RUnlock()
+
+	if !n.nopts.SnapshotEnabled() {
+		n.metrics.snapshot.ReportSuccess(n.nowFn().Sub(callStart))
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.GetOwnedShards()
+	for _, shard := range shards {
+		isSnapshotting, lastSuccessfulSnapshot := shard.SnapshotState()
+		if isSnapshotting {
+			// Should never happen because snapshots should never overlap
+			// each other (controlled by loop in flush manager)
+			n.log.
```

Comment:
formatting nit, putting withFields on the same line is probably cleaner

Reply:
tried, looks a little cluttered IMO

Reply:
cool